autoload/clap/provider/git_files.vim

```diff
@@ -6,14 +6,32 @@ set cpoptions&vim
 
 let s:git_files = {}
 
-if executable('git')
-  let s:git_files.source = 'git ls-files '.(has('win32') ? '' : ' | uniq')
-else
-  let s:git_files.source = ['git executable not found']
-endif
+function! s:git_files.source() abort
+  if !executable('git')
+    call clap#helper#echo_error('git executable not found')
+    return []
+  endif
+
+  let changed = systemlist('git ls-files '.(has('win32') ? '' : ' | uniq'))
```

Comment:
Simply changing the `source` to a function returning a List will make this provider always run in sync way. It's ok in most cases, but it will be a problem when the git files are too many, this system call could be slow, making vim-clap sluggish and you'll complain that.

Check out `converter` for adding an icon using VimL, https://github.com/liuchengxu/vim-clap#create-pure-async-provider.

I'll see why `git_files` returns nothing.

Reply:
Ok. Would you suggest that a pure async provider would be better, then?

Reply:
Hmm. I would guess not, since pure async providers don't seem to have a way to specify a source to run without a query (i.e. when `on_typed` doesn't really make sense, as here).

Reply:
Actually the non-pure async provider can also have the `converter` property, I forget to document that. But it's better to provide the icon support in the external Rust extension `maple` since we already use it.

Reply:
Ok. I looked into that briefly, by modifying https://github.com/liuchengxu/vim-clap/blob/master/autoload/clap/maple.vim#L124 to also test against `git_files`, but it didn't add icons when I ran it. It seems that there are a number of places in the code with special-case handling for the providers that support icons; maybe it would make sense to refactor `maple` icon support to be a property of a provider, then just test if this property is set?

Reply:
Adding an icon support needs both `source` and `sink` to be tweaked. If there is a clear pattern, we could make it easier by introducing some property into the provider.